### PR TITLE
Add functionality to rerun failed tests

### LINKED ISSUE #24642: Support re-running failed tests
**Description:**

The test framework should support re-running failed test cases instead of running the entire suite. This saves time and resources spent for troubleshooting flaky tests. All the failed and skipped tests should be run when attempted to re-run.

### PR DESCRIPTION
**Pull request by Ibaqu**
## Purpose
Implement Rerun failed test cases functionality

Fixes #24642

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: cli/ballerina-packerina/src/main/java/org/ballerinalang/packerina/cmd/TestCommand.java [hunk 1]
```diff
@@ -159,6 +159,9 @@ public TestCommand(Path userDir, PrintStream outStream, PrintStream errStream, b
     @CommandLine.Option(names = "--old-parser", description = "Enable old parser.", hidden = true)
     private boolean useOldParser;
 
+    @CommandLine.Option(names = "--rerun-failedTests", description = "Rerun failed tests.")
```
Comment (azinneera) [inline cli/ballerina-packerina/src/main/java/org/ballerinalang/packerina/cmd/TestCommand.java:162]: Shall we rename the flag to `--rerun-failed`?

Comment (Ibaqu) [reply cli/ballerina-packerina/src/main/java/org/ballerinalang/packerina/cmd/TestCommand.java:162]: Ack. Will change to `--rerun-failed`

File: cli/ballerina-packerina/src/main/java/org/ballerinalang/packerina/task/CleanTargetDirTask.java [hunk 2]
```diff
@@ -58,8 +59,14 @@ public void execute(BuildContext buildContext) {
                         + File.separator + ProjectDirConstants.BIN_DIR_NAME).toPath());
                 deleteResource(new File(targetDir.toAbsolutePath().toString()
                         + File.separator + ProjectDirConstants.TARGET_BALO_DIRECTORY).toPath());
-                deleteResource(new File(targetDir.toAbsolutePath().toString()
-                        + File.separator + ProjectDirConstants.CACHES_DIR_NAME).toPath());
+//                deleteResource(new File(targetDir.toAbsolutePath().toString()
```
Comment (azinneera) [inline cli/ballerina-packerina/src/main/java/org/ballerinalang/packerina/task/CleanTargetDirTask.java:62]: Can these comments can be removed?

Comment (Ibaqu) [reply cli/ballerina-packerina/src/main/java/org/ballerinalang/packerina/task/CleanTargetDirTask.java]: removed changes to CleanTargetDirTask

File: misc/testerina/modules/testerina-runtime/src/main/java/org/ballerinalang/test/runtime/BTestRunner.java [hunk 3]
```diff
@@ -404,6 +413,13 @@ private void executeFunction(Test test, TestSuite suite, String packageName, Cla
                                  Scheduler scheduler, AtomicBoolean shouldSkip, AtomicBoolean shouldSkipTest,
                                  List<String> failedOrSkippedTests, List<String> failedAfterFuncTests) {
         TesterinaResult functionResult;
+        Path sourceRootPath = Paths.get(suite.getSourceRootPath());
+        Path jsonCacheDir = sourceRootPath.resolve("target").resolve("caches").resolve("json_cache");
+        Path rerunJson =
+                jsonCacheDir.resolve(suite.getOrgName()).resolve(suite.getPackageID()).resolve(suite.getVersion());
+        Path jsonPath = Paths.get(rerunJson.toString(), "retun_test.json");
```
Comment (Dilhasha) [inline misc/testerina/modules/testerina-runtime/src/main/java/org/ballerinalang/test/runtime/BTestRunner.java:420]: Shall we correct the typo in "retun_test.json"

Comment (Ibaqu) [reply misc/testerina/modules/testerina-runtime/src/main/java/org/ballerinalang/test/runtime/BTestRunner.java]: Corrected typo in all instances.

File: cli/ballerina-packerina/src/main/java/org/ballerinalang/packerina/cmd/TestCommand.java [hunk 4]
```diff
@@ -204,6 +207,28 @@ public void execute() {
             return;
         }
 
+        if (this.rerunTests) {
+            // Cannot rerun failed tests using -a | -all flags
+            if (this.buildAll) {
+                CommandUtil.printError(this.errStream,
+                                       "Cannot specify --rerun-failed and -a | --all flags at the same time",
```
Comment (azinneera) [inline cli/ballerina-packerina/src/main/java/org/ballerinalang/packerina/cmd/TestCommand.java:214]: Once the new package implementation is in place, we should allow running with the `-a` flag. Please create an issue to track this.

**Review by Dilhasha — APPROVED** [review]
LGTM

File: cli/ballerina-packerina/src/main/java/org/ballerinalang/packerina/cmd/TestCommand.java (hunk 4 above)
Comment (Ibaqu) [reply cli/ballerina-packerina/src/main/java/org/ballerinalang/packerina/cmd/TestCommand.java:214]: Ack. Created issue to track this : https://github.com/ballerina-platform/ballerina-lang/issues/25382

File: cli/ballerina-packerina/src/main/java/org/ballerinalang/packerina/task/RunTestsTask.java [hunk 5]
```diff
@@ -432,4 +458,11 @@ private ModuleStatus loadModuleStatusFromFile(Path statusJsonPath) throws IOExce
         BufferedReader bufferedReader = Files.newBufferedReader(statusJsonPath, StandardCharsets.UTF_8);
         return gson.fromJson(bufferedReader, ModuleStatus.class);
     }
+
+    private List<String> readFailedTestsFromFile(Path rerunTestJsonPath) throws IOException {
+        Gson gson = new Gson();
+        rerunTestJsonPath = Paths.get(rerunTestJsonPath.toString(), "rerun_test.json");
+        BufferedReader bufferedReader = Files.newBufferedReader(rerunTestJsonPath, StandardCharsets.UTF_8);
```
Comment (azinneera) [inline cli/ballerina-packerina/src/main/java/org/ballerinalang/packerina/task/RunTestsTask.java:465]: We should catch the exception and close the `BufferedReader`. Lets wrap this with a try-with-resources block.

File: tests/testerina-integration-test/src/test/java/org/ballerinalang/testerina/test/RerunFailedTest.java [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+        String msg2 = "2 failing";
+        LogLeecher clientLeecher1 = new LogLeecher(msg1);
+        LogLeecher clientLeecher2 = new LogLeecher(msg2);
+
+        balClient.runMain("test", new String[]{"--rerun-failed", "module_1"}, null, new String[]{},
+                new LogLeecher[]{clientLeecher1, clientLeecher2}, projectPath);
+
+        clientLeecher1.waitForText(20000);
+        clientLeecher2.waitForText(20000);
+    }
```
Comment (azinneera) [inline tests/testerina-integration-test/src/test/java/org/ballerinalang/testerina/test/RerunFailedTest.java:67]: Lets add a test case to verify the behavior when there are are no failed tests found.

Comment (Ibaqu) [reply tests/testerina-integration-test/src/test/java/org/ballerinalang/testerina/test/RerunFailedTest.java:67]: Added `testRerunWithNoFailedTests`

File: cli/ballerina-packerina/src/main/java/org/ballerinalang/packerina/task/RunTestsTask.java (hunk 5 above)
Comment (Ibaqu) [reply cli/ballerina-packerina/src/main/java/org/ballerinalang/packerina/task/RunTestsTask.java]: Used `try-with-resource` block and handled exceptions within the function itself

**Review by azinneera — APPROVED** [review]

**Review by Dilhasha — APPROVED** [review]

**Review by Dilhasha — APPROVED** [review]

**Pull request merged**